TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/BlockLibrary/SamplerStatePropertyBlockBuilder.cs

```diff
@@ -0,0 +1,79 @@
+
+        public Block Build(ShaderContainer container)
+        {
+            var attributes = BuildAttributes(container);
+            return BuildWithAttributeOverrides(container, attributes);
+        }
+
+        // Builds with the provided attributes.
+        // This primarily used for unit testing invalid attribute cases.
+        public Block BuildWithAttributeOverrides(ShaderContainer container, List<ShaderAttribute> attributes)
+        {
+            var propData = new BlockBuilderUtilities.PropertyDeclarationData
+            {
+                Fields = new List<BlockBuilderUtilities.FieldData>
+                {
+                    // Build the sampler state property
+                    new BlockBuilderUtilities.FieldData
+                    {
+                        Name = FieldName,
+                        Type = container._UnitySamplerState,
+                        PropertyAttribute = PropertyAttribute,
+                        ExtraAttributes = attributes,
+                    },
+                    // Build a texture property that we can sample. This would be needed to actually use the sampler in a play mode test
+                    new BlockBuilderUtilities.FieldData
+                    {
+                        Name = "_Texture",
+                        Type = container._UnityTexture2D,
+                        PropertyAttribute = new PropertyAttributeData(),
+                    },
+                },
+                OutputsAssignmentCallback = (builder, propData) =>
+                {
+                    builder.AddLine($"float2 uv = float2(0, 0);");
+                    builder.AddLine($"float4 sample = inputs._Texture.Sample(inputs.{FieldName}.samplerstate, uv);");
+                    builder.AddLine($"outputs.BaseColor = sample.xyz;");
+                    builder.AddLine($"outputs.Alpha = sample.w;");
+                }
+            };
+            return BlockBuilderUtilities.CreateSimplePropertyBlock(container, BlockName, propData);
+        }
+
+        internal List<ShaderAttribute> BuildAttributes(ShaderContainer container)
+        {
+            var attributes = new List<ShaderAttribute>();
+            var attributeBuilder = new ShaderAttribute.Builder(container, SamplerStateAttribute.AttributeName);
+            if (FilterMode is SamplerStateAttribute.FilterModeEnum filterMode)
+                attributeBuilder.Param(SamplerStateAttribute.FilterModeParamName, filterMode.ToString());
+            var wrapModeString = SamplerStateAttribute.BuildWrapModeParameterValue(WrapModes);
+            if (!string.IsNullOrEmpty(wrapModeString))
+                attributeBuilder.Param(SamplerStateAttribute.WrapModeParamName, wrapModeString);
+            if (DepthCompare is bool depthCompare)
+                attributeBuilder.Param(SamplerStateAttribute.DepthCompareParamName, depthCompare.ToString());
+            if (AnisotropicLevel is int anisotropicLevel)
```

Comment:
Note: we support only a fixed list of anise filters (2x, 4x, 8x and 16x).
Though the compiler checks for validity, perhaps we should use an enum here.

Reply:
Good idea, I like the enum over the integer

Reply:
So this is actually the test code, I do the validity checks already. I'm not sure about switching to an enum in the attribute though as the naming is harder (you can't have an enum value that starts with a number).
Currently it is:
`[SamplerState(anisotropicLevel = 2)]`

Reply:
The validation is here btw [anisotropicLevel](https://github.com/Unity-Technologies/Graphics/pull/7334/files#diff-7c42bde1d3f45ca0061f19267815af4c41a795b1c0884f1069ba8f11424b8717R133)

Reply:
What about AnisoLevel2X etc?